### PR TITLE
Fix gcc13 warnings in RDF

### PR DESCRIPTION
**Pull request by linev**

### CODE REVIEW

File: tree/dataframe/inc/ROOT/RDF/RInterfaceBase.hxx [hunk 1]
```diff
@@ -108,9 +108,7 @@ protected:
          const auto &retTypeID = typeid(typename RetType::value_type);
          const auto &colName = colNames[0]; // we have only one element in there
          const auto *define = fColRegister.GetDefine(colName);
-         const auto &expectedTypeID =
-            define ? define->GetTypeId() : RDFInternal::TypeName2TypeID(GetColumnType(colName));
-         if (retTypeID != expectedTypeID)
+         if (define ? (retTypeID != define->GetTypeId()) : (retTypeID != RDFInternal::TypeName2TypeID(GetColumnType(colName))))
```
Comment (jalopezg-git) [inline tree/dataframe/inc/ROOT/RDF/RInterfaceBase.hxx:111]: I find the previous code much more readable.  Could you paste here the compiler diagnostic from gcc13 to see if we can find a better fix?

**Comment by linev** [comment]
Here full output:
```
In file included from /home/linev/git/webgui/tree/dataframe/inc/ROOT/RDF/RInterface.hxx:22,
                 from /home/linev/git/webgui/tree/dataframe/inc/ROOT/RDataFrame.hxx:21,
                 from /home/linev/git/webgui/tree/dataframe/test/dataframe_datasetspec.cxx:8:
/home/linev/git/webgui/tree/dataframe/inc/ROOT/RDF/RInterfaceBase.hxx: In instantiation of ‘void ROOT::RDF::RInterfaceBase::SanityChecksForVary(const std::vector<std::__cxx11::basic_string<char> >&, const std::vector<std::__cxx11::basic_string<char> >&, std::string_view) [with RetType = ROOT::VecOps::RVec<long long unsigned int>; std::string_view = std::basic_string_view<char>]’:
/home/linev/git/webgui/tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2841:35:   required from ‘ROOT::RDF::RInterface<T, V> ROOT::RDF::RInterface<T, V>::VaryImpl(const std::vector<std::__cxx11::basic_string<char> >&, F&&, const ROOT::RDF::ColumnNames_t&, const std::vector<std::__cxx11::basic_string<char> >&, std::string_view) [with bool IsSingleColumn = true; F = RDatasetSpecTest_FilterDependingOnVariation_Test::TestBody()::<lambda(ULong64_t)>; Proxied = ROOT::Detail::RDF::RLoopManager; DataSource = void; ROOT::RDF::ColumnNames_t = std::vector<std::__cxx11::basic_string<char> >; std::string_view = std::basic_string_view<char>]’
/home/linev/git/webgui/tree/dataframe/inc/ROOT/RDF/RInterface.hxx:676:28:   required from ‘ROOT::RDF::RInterface<T, V> ROOT::RDF::RInterface<T, V>::Vary(std::string_view, F&&, const ROOT::RDF::ColumnNames_t&, const std::vector<std::__cxx11::basic_string<char> >&, std::string_view) [with F = RDatasetSpecTest_FilterDependingOnVariation_Test::TestBody()::<lambda(ULong64_t)>; Proxied = ROOT::Detail::RDF::RLoopManager; DataSource = void; std::string_view = std::basic_string_view<char>; ROOT::RDF::ColumnNames_t = std::vector<std::__cxx11::basic_string<char> >]’
/home/linev/git/webgui/tree/dataframe/inc/ROOT/RDF/RInterface.hxx:697:18:   required from ‘ROOT::RDF::RInterface<T, V> ROOT::RDF::RInterface<T, V>::Vary(std::string_view, F&&, const ROOT::RDF::ColumnNames_t&, std::size_t, std::string_view) [with F = RDatasetSpecTest_FilterDependingOnVariation_Test::TestBody()::<lambda(ULong64_t)>; Proxied = ROOT::Detail::RDF::RLoopManager; DataSource = void; std::string_view = std::basic_string_view<char>; ROOT::RDF::ColumnNames_t = std::vector<std::__cxx11::basic_string<char> >; std::size_t = long unsigned int]’
/home/linev/git/webgui/tree/dataframe/test/dataframe_datasetspec.cxx:441:22:   required from here
/home/linev/git/webgui/tree/dataframe/inc/ROOT/RDF/RInterfaceBase.hxx:111:22: error: possibly dangling reference to a temporary [-Werror=dangling-reference]
  111 |          const auto &expectedTypeID =
      |                      ^~~~~~~~~~~~~~
/home/linev/git/webgui/tree/dataframe/inc/ROOT/RDF/RInterfaceBase.hxx:112:72: note: the temporary was destroyed at the end of the full expression ‘ROOT::Internal::RDF::TypeName2TypeID(ROOT::RDF::RInterfaceBase::GetColumnType(std::string_view)((& colName)->std::__cxx11::basic_string<char>::operator __sv_type()))’
  112 |             define ? define->GetTypeId() : RDFInternal::TypeName2TypeID(GetColumnType(colName));
      |                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by linev** [comment]
No idea. 
For me return type in both cases looked ok - ` const std::type_info &`.
Only a question how conditional operator `?` implemented in gcc13 - may be it makes a problem.

**Comment by linev** [comment]
Another idea.
May be `gcc` looks into implementation of `TypeName2TypeID` and see that for some cases it creates exception instead of returning value.

**Comment by jalopezg-git** [comment]
> Hi, I don't understand what temporary gcc's message refers to?
> 
>     * `GetColumnType` returns a `string` by value that is taken in as a `const string &` by `TypeName2TypeID`, so everything should be fine there
> 
>     * `TypeName2TypeID` returns a `const type_info &` that is conditionally assigned to a `const auto &` and that should also be ok: the `const type_info &` is produced by expressions such as `typeid(bool)` and that value should have static lifetime
> 
> 
> Help?

GCC13 has not been officially released yet.  It comes with the [new `-Wdangling-reference` diagnostic](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=106393), which I think still provides many false positives, e.g. [see this](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=107532).

In particular, this warning seems absolutely bogus.  It can be reproduced with this simple excerpt (compile with `g++ -Werror=dangling-reference`).
```c++
#include <string>
#include <typeinfo>

const std::type_info &f(const std::string &);
void g() {
   // A `std::string` temporary is implicitly constructed passing the string literal "foo" to the ctor.  Apparently,
   // this triggers one of the patterns in GCC's new `[-Wdangling-reference]`. 
   const auto &ti = f("foo");
}
```

IMO, this warning is totally bogus (and might be seen in other parts of ROOT).  Thus, I would suggest not changing the code but conditionally ignoring it,
_(a)_ Either in this translation unit only
```c++
#pragma GCC diagnostic ignored ...
```
, or _(b)_ most likely for all ROOT (appending `-Wno-dangling-reference` from CMake).

I would go for `_(b)_`, at least until the feature becomes mature enough and does not yield false positives.  What do you think, @linev, @eguiraud?

Also CC: @Axel-Naumann and @hahnjo.

**Comment by Axel-Naumann** [comment]
I'd prefer to only act after the release of GCC13, and until then address the correct diagnostics so we are left with only false positives.

**Comment by linev** [comment]
Then I propose to left only first commit - which about automatic conversion of array of `const char*` into `vector<string>`
Second commit can be ignored.

File: tree/dataframe/inc/ROOT/RDF/RResultMap.hxx [hunk 2]
```diff
@@ -83,7 +83,7 @@ class RResultMap {
               std::vector<std::string> &&keys, ROOT::Detail::RDF::RLoopManager &lm,
               std::shared_ptr<ROOT::Internal::RDF::RActionBase> nominalAction,
               std::shared_ptr<ROOT::Internal::RDF::RActionBase> variedAction)
-      : fKeys{ROOT::Internal::RDF::Union({"nominal"}, keys)}, fLoopManager(&lm),
+      : fKeys{ROOT::Internal::RDF::Union({std::string("nominal")}, keys)}, fLoopManager(&lm),
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RResultMap.hxx:86]:
```suggestion
      : fKeys{ROOT::Internal::RDF::Union({"nominal"s}, keys)}, fLoopManager(&lm),
```
would be cleaner but then we'd need a `using namespace std::string_literals` somewhere. as you prefer.

**Review by eguiraud — APPROVED** [review]
Thank you @linev !

**Pull request merged**